src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -89,18 +90,40 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function compile()
     {
+        if ($this->isCompiled()) {
+            return;
+        }
+
         $this->parameterBag->resolve();
 
-        $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        if (!$this->parameterBag instanceof FrozenParameterBag) {
```

Comment:
This should be removed: changing the object instance of the bag can be seen as a wanted side effect of compiling. Not wrapping a FrozenParameterBag in a FrozenParameterBag looks like an unnecessary optimization to me also.